### PR TITLE
Improve Prime:prime? algorithm up to 4 times faster

### PR DESCRIPTION
**Pull request by alexanderluna**
**Add a faster Prime:prime? algorithm**
I worked the other day on a problem from the Euler Project and found myself using Prime:prime? to iterate over a large set of numbers. While doing that I noticed that the Prime:prime? method was a little slow. I swapped it with another algorithm I found on [SO](https://stackoverflow.com/questions/1801391/what-is-the-best-algorithm-for-checking-if-a-number-is-prime). The result was incredible:
```bash
$ (Prime:prime?) Time to complete (10 million): 205.742023
$ (Prime:prime?) 664579 prime numbers found

$ (new_prime?) Time to complete (10 million): 50.634685
$ (new_prime?) 664579 prime numbers found
```
As you can see iterating over 10 million numbers, the new algorithm is able to find the same prime numbers 4 times faster. I ran the tests locally and everything seems to work fine.
Could you double check ?

### CODE REVIEW

**Comment by Mycroft1891** [comment]
sorry, I ran the `spec` tests at first and didn't notice the `test` folder which is why I had to push 2 fixes.

File: lib/prime.rb [hunk 1]
```diff
@@ -32,15 +32,17 @@ def prime_division(generator = Prime::Generator23.new)
...
+    return false if self % 2 == 0 || self % 3 == 0
+    i = 5
+    w = 2
+    while i * i <= self
+        return false if self % i == 0
+        i += w
+        w = 6 - w
     end
-    true
+    return true
```
Comment (yuki24) [inline lib/prime.rb]: No need to declare `return`.

Comment (Mycroft1891) [reply lib/prime.rb]: you are right

**Comment by tom-lord** [comment]
I ran the proposed method, and did see a *slight* (20%) improvement, but no where near the 400% speedup claimed. I'd be curious to see the actual test ran, rather than just the result @Mycroft1891 ?

Here is my *complete* benchmarking code, with results:

```
class Integer
  def prime_orig?
    return self >= 2 if self <= 3
    return true if self == 5
    return false unless 30.gcd(self) == 1
    (7..Integer.sqrt(self)).step(30) do |p|
      return false if
        self%(p)    == 0 || self%(p+4)  == 0 || self%(p+6)  == 0 || self%(p+10) == 0 ||
        self%(p+12) == 0 || self%(p+16) == 0 || self%(p+22) == 0 || self%(p+24) == 0
    end
    true
  end

  def prime_new?
    return false if self <= 1
    return true  if self == 2 || self == 3
    return false if self % 2 == 0 || self % 3 == 0
    i = 5
    w = 2
    while i * i <= self
      return false if self % i == 0
      i += w
      w = 6 - w
    end
    true
  end
end

require 'benchmark'
iterations = 100_000
numbers = (-100..1_000).to_a

Benchmark.bmbm(20) do |bm|
  bm.report('Original method') do
    iterations.times do
      numbers.select(&:prime_orig?)
    end
  end

  bm.report('New method') do
    iterations.times do
      numbers.select(&:prime_new?)
    end
  end
end

Rehearsal --------------------------------------------------------
Original method       31.264767   0.097343  31.362110 ( 31.447996)
New method            24.547395   0.131066  24.678461 ( 24.687701)
---------------------------------------------- total: 56.040571sec

                           user     system      total        real
Original method       31.111096   0.040542  31.151638 ( 31.177937)
New method            24.643408   0.160427  24.803835 ( 24.895586)
```

Note that real world usage of `Integer#prime?` may differ from the above test case, so this is by no means a definitive result.

**Comment by Mycroft1891** [comment]
My test was much simpler, I will try out yours a little later.
For now here is my test:

```
require('prime')
class Integer

  def prime_new?
    return false if self <= 1
    return true  if self == 2 || self == 3
    return false if self % 2 == 0 || self % 3 == 0
    i = 5
    w = 2
    while i * i <= self
      return false if self % i == 0
      i += w
      w = 6 - w
    end
    true
  end
end


ruby_prime = []
my_prime = []

t = Time.now
10_000_000.times do |i|
  ruby_prime.push(i) if Prime.prime?(i)
end
puts "(Prime) Time to complete (100k): #{Time.now - t}"
puts "(Prime) #{ruby_prime.count} prime numbers found\n\n"

b = Time.now
10_000_000.times do |i|
  my_prime.push(i) if i.isprime?
end

puts "(isprime?) Time to complete (100k): #{Time.now - b}"
puts "(isprime?) #{my_prime.count} prime numbers found\n"
```

As you increase the number of times, the time difference becomes greater. I put the integer class extension in a different file and imported it like Prime actually at the top to make it a little more fair.

**Comment by yuki24** [comment]
@Mycroft1891 Could you re-write the script to use `Benchmark` or [`benchmark-ips`](https://github.com/evanphx/benchmark-ips) like @tom-lord did? Also it would be great if you could share comparisons for multiple cases (e.g. each for negative number, 0, 1, 2, 3, small prime number, small non-prime number, large prime number, large non-prime number, etc)

**Comment by Mycroft1891** [comment]
@yuki24 Sure here is the same test using Benchmark:

```
require 'benchmark'
require 'prime'
require './my_prime'

iterations = 1_000_000

ruby_prime = []
my_prime = []

Benchmark.bm do |bm|
  bm.report do
    iterations.times do |i|
      ruby_prime.push(i) if Prime.prime?(i)
    end
  end

  bm.report do
    iterations.times do |i|
      my_prime.push(i) if isprime?(i)
    end
  end
end
```

**Comment by yuki24** [comment]
@Mycroft1891 a few more comments:

 * Inline the `./my_prime` so anyone can your script
 * Remove the use of the `push` method. It'd just add some overhead to the benchmark result
 * As I mentioned above, benchmark more cases. At least with a negative number, 0, 1, 2, 3, small prime number, small non-prime number, large prime number, large non-prime number. The `prime?` method has a lot of conditionals and it is important to understand what's significantly faster and what's slightly faster.

**Comment by Mycroft1891** [comment]
@yuki24 @tom-lord I found out why I saw such a big difference. If you run the test using `Prime.prime?(n)` the performance drops by a lot while when you use `n.prime?` it is much faster as @tom-lord noted.
I created a `my_prime.rb` file this is inside it:

```
# my_prime.rb
class Prime
  def isprime?(n)
    return false if n <= 1
    return true  if n == 2 || n == 3
    return false if n % 2 == 0 || n % 3 == 0
    i = 5
    w = 2
    while i * i <= n
      return false if n % i == 0
      i += w
      w = 6 - w
    end
    true
  end
end

class Integer
  def isprime?
    return false if self <= 1
    return true  if self == 2 || self == 3
    return false if self % 2 == 0 || self % 3 == 0
    i = 5
    w = 2
    while i * i <= self
        return false if self % i == 0
        i += w
        w = 6 - w
    end
    true
  end
end

```

As you can see I extended both the Integer and the Prime class just to make it similar.

I created a new benchmark covering:
- negative number (-100..1000)
- small prime numbers (first 2000)
- small non prime numbers (any non prime smaller than 1000)
- large prime numbers (primes in this range 10_999_000..11_000_000)
- large non prime numbers (non primes in this range 10_999_000..11_000_000)

In the first test, I run it comparing `n.prime?` vs. `n.isprime?` the second is the new prime algorithm.

```
require 'benchmark'
require 'prime'
require './my_prime'

one_million      = 1_000_000
ten_thousand     = 10_000
thousand         = 1_000

with_negative    = (-100..1_000).to_a
small_non_primes = (0..1_000).reject {|n| Prime.prime?(n) }
small_primes     = Prime.take(2000)
large_non_primes = (10_999_000..11_000_000).reject {|n| Prime.prime?(n) }
large_primes     = (10_999_000..11_000_000).select {|n| Prime.prime?(n) }

Benchmark.bmbm(20) do |bm|
  bm.report('Old: 1_000_000 * prime?') do
    one_million.times do |i|
      i.prime?
    end
  end

  bm.report('New: 1_000_000 * prime?') do
    one_million.times do |i|
      i.isprime?
    end
  end

  bm.report('Old: 10_000 * -100 to 1_000') do
    ten_thousand.times do |i|
      with_negative.select(&:prime?)
    end
  end

  bm.report('New: 10_000 * -100 to 1_000') do
    ten_thousand.times do |i|
      with_negative.select(&:isprime?)
    end
  end

  bm.report('Old: 1_000 * small non primes') do
    thousand.times do |i|
      small_non_primes.select(&:prime?)
    end
  end

  bm.report('New: 1_000 * small non primes') do
    thousand.times do |i|
      small_non_primes.select(&:isprime?)
    end
  end

  bm.report('Old: 1_000 * small primes') do
    thousand.times do |i|
      small_primes.select(&:prime?)
    end
  end

  bm.report('New: 1_000 * small primes') do
    thousand.times do |i|
      small_primes.select(&:isprime?)
    end
  end

  bm.report('Old: 1_000 * large primes') do
    thousand.times do |i|
      large_primes.select(&:prime?)
    end
  end

  bm.report('New: 1_000 * large primes') do
    thousand.times do |i|
      large_primes.select(&:isprime?)
    end
  end

  bm.report('Old: 1_000 * large non primes') do
    thousand.times do |i|
      large_non_primes.select(&:prime?)
    end
  end

  bm.report('New: 1_000 * large non primes') do
    thousand.times do |i|
      large_non_primes.select(&:isprime?)
    end
  end
end
```
The results are almost the same, sometimes the new algorithm is faster while most of the times the old one is faster: 

```
Rehearsal -----------------------------------------------------------------
Old: 1_000_000 * prime?         1.405529   0.007792   1.413321 (  1.428931)
New: 1_000_000 * prime?         2.073816   0.012818   2.086634 (  2.131261)
Old: 10_000 * -100 to 1_000     4.592670   0.025376   4.618046 (  4.672404)
New: 10_000 * -100 to 1_000     3.636590   0.052335   3.688925 (  3.743397)
Old: 1_000 * small non primes   0.286702   0.002224   0.288926 (  0.291118)
New: 1_000 * small non primes   0.204080   0.001687   0.205767 (  0.207332)
Old: 1_000 * small primes       3.848818   0.022083   3.870901 (  3.911539)
New: 1_000 * small primes       5.210528   0.035932   5.246460 (  5.318403)
Old: 1_000 * large primes       3.175398   0.018349   3.193747 (  3.244483)
New: 1_000 * large primes       5.894557   0.031974   5.926531 (  5.989830)
Old: 1_000 * large non primes   1.273114   0.007360   1.280474 (  1.298590)
New: 1_000 * large non primes   1.761014   0.010536   1.771550 (  1.803974)
------------------------------------------------------- total: 33.591282sec

                                    user     system      total        real
Old: 1_000_000 * prime?         1.442458   0.008244   1.450702 (  1.471786)
New: 1_000_000 * prime?         2.097896   0.012572   2.110468 (  2.142132)
Old: 10_000 * -100 to 1_000     4.709277   0.026490   4.735767 (  4.794737)
New: 10_000 * -100 to 1_000     3.574214   0.027443   3.601657 (  3.657275)
Old: 1_000 * small non primes   0.279802   0.002282   0.282084 (  0.286573)
New: 1_000 * small non primes   0.222431   0.001646   0.224077 (  0.227809)
Old: 1_000 * small primes       3.824703   0.024839   3.849542 (  3.891357)
New: 1_000 * small primes       5.274420   0.030545   5.304965 (  5.372510)
Old: 1_000 * large primes       3.062497   0.016671   3.079168 (  3.110922)
New: 1_000 * large primes       5.765497   0.030927   5.796424 (  5.870993)
Old: 1_000 * large non primes   1.225203   0.007635   1.232838 (  1.254354)
New: 1_000 * large non primes   1.758376   0.008997   1.767373 (  1.790090)
```

Everything changes when you compare `Prime.prime?(n)` vs. `Prime.isprime?(n)` once again the second one is the new prime method.

```
require 'benchmark'
require 'prime'
require './my_prime'

one_million      = 1_000_000
ten_thousand     = 10_000
thousand         = 1_000

with_negative    = (-100..1_000).to_a
small_non_primes = (0..1_000).reject {|n| Prime.prime?(n) }
small_primes     = Prime.take(2000)
large_non_primes = (10_999_000..11_000_000).reject {|n| Prime.prime?(n) }
large_primes     = (10_999_000..11_000_000).select {|n| Prime.prime?(n) }

Benchmark.bmbm(20) do |bm|
  bm.report('Old: 1_000_000 * prime?') do
    one_million.times do |i|
      Prime.prime?(i)
    end
  end

  bm.report('New: 1_000_000 * prime?') do
    one_million.times do |i|
      i.isprime?
    end
  end

  bm.report('Old: 10_000 * -100 to 1_000') do
    ten_thousand.times do |i|
      with_negative.select { |n| Prime.prime?(n) }
    end
  end

  bm.report('New: 10_000 * -100 to 1_000') do
    ten_thousand.times do |i|
      with_negative.select { |n| Prime.isprime?(n) }
    end
  end

  bm.report('Old: 1_000 * small non primes') do
    thousand.times do |i|
      small_non_primes.select { |n| Prime.prime?(n) }
    end
  end

  bm.report('New: 1_000 * small non primes') do
    thousand.times do |i|
      small_non_primes.select { |n| Prime.isprime?(n) }
    end
  end

  bm.report('Old: 1_000 * small primes') do
    thousand.times do |i|
      small_primes.select { |n| Prime.prime?(n) }
    end
  end

  bm.report('New: 1_000 * small primes') do
    thousand.times do |i|
      small_primes.select { |n| Prime.isprime?(n) }
    end
  end

  bm.report('Old: 1_000 * large primes') do
    thousand.times do |i|
      large_primes.select { |n| Prime.prime?(n) }
    end
  end

  bm.report('New: 1_000 * large primes') do
    thousand.times do |i|
      large_primes.select { |n| Prime.isprime?(n) }
    end
  end

  bm.report('Old: 1_000 * large non primes') do
    thousand.times do |i|
      large_non_primes.select { |n| Prime.prime?(n) }
    end
  end

  bm.report('New: 1_000 * large non primes') do
    thousand.times do |i|
      large_non_primes.select { |n| Prime.isprime?(n) }
    end
  end
end
```

These are the results I saw when I tested it:

```
Rehearsal -----------------------------------------------------------------
Old: 1_000_000 * prime?         9.707928   0.056095   9.764023 (  9.884189)
New: 1_000_000 * prime?         2.068627   0.005823   2.074450 (  2.088193)
Old: 10_000 * -100 to 1_000    27.125947   0.105004  27.230951 ( 27.481636)
New: 10_000 * -100 to 1_000     7.366718   0.038283   7.405001 (  7.492338)
Old: 1_000 * small non primes   1.822184   0.013414   1.835598 (  1.863551)
New: 1_000 * small non primes   0.495780   0.002681   0.498461 (  0.503347)
Old: 1_000 * small primes      23.748546   0.115993  23.864539 ( 24.070753)
New: 1_000 * small primes       6.046003   0.032608   6.078611 (  6.144513)
Old: 1_000 * large primes      23.265469   0.096873  23.362342 ( 23.591089)
New: 1_000 * large primes       5.896811   0.034694   5.931505 (  6.014809)
Old: 1_000 * large non primes   8.249851   0.045913   8.295764 (  8.376780)
New: 1_000 * large non primes   2.080070   0.011458   2.091528 (  2.110891)
------------------------------------------------------ total: 118.432773sec

                                    user     system      total        real
Old: 1_000_000 * prime?         9.740492   0.054909   9.795401 (  9.903361)
New: 1_000_000 * prime?         2.052631   0.011364   2.063995 (  2.087338)
Old: 10_000 * -100 to 1_000    26.913940   0.099397  27.013337 ( 27.217281)
New: 10_000 * -100 to 1_000     7.545237   0.044933   7.590170 (  7.685570)
Old: 1_000 * small non primes   1.792154   0.010089   1.802243 (  1.821959)
New: 1_000 * small non primes   0.492120   0.009264   0.501384 (  0.511246)
Old: 1_000 * small primes      23.592904   0.096804  23.689708 ( 23.948073)
New: 1_000 * small primes       5.971981   0.033575   6.005556 (  6.079808)
Old: 1_000 * large primes      23.149618   0.102074  23.251692 ( 23.496745)
New: 1_000 * large primes       5.896328   0.029026   5.925354 (  5.996483)
Old: 1_000 * large non primes   8.138648   0.046656   8.185304 (  8.269999)
New: 1_000 * large non primes   2.098734   0.011617   2.110351 (  2.137368)
```

Basically `Integer.prime?` is much much faster than `Prime.prime?(n)` 
以上です

**Comment by tom-lord** [comment]
So the original comparison was between apples and oranges; you'd compared `Prime.prime?` with `Integer#prime?` 😄
 
(I thought it was a bit of odd that this PR is titled `Improve Prime:prime? ...` when you didn't actually touch that method!)

Based on those performance figures, I don't think there's sufficient justification for changing the implementation. For larger values (where performance is more critical!), the new method performs significantly slower.

**Comment by Mycroft1891** [comment]
@tom-lord Yes, my error. `Integer.prime?` and `Prime.prime?(n)` confused me thus I changed the wrong method. My only question is why do we have 2 ways of checking primes ? I couldn't find in the docs anything about why `Prime.prime?(n)` is so much slower except that it takes an optional generator. Wouldn't it be better to point our that is performs much worse than `Integer.prime?`

**Comment by k0kubun** [comment]
It seems to have a conflict now. Could you rebase this from master?
